xds/internal/client/filter_chain.go

```diff
@@ -0,0 +1,570 @@
+	}
+	return matchingDstPrefixes
+}
+
+// filterBasedOnSourceType is the second stage of the matching algorithm. It
+// trims the filter chains based on the most specific source type match.
+func filterBasedOnSourceType(dstPrefixes []*destPrefixEntry, srcType SourceType) []*sourcePrefixes {
+	var (
+		srcPrefixes      []*sourcePrefixes
+		bestSrcTypeMatch int
+	)
+	for _, prefix := range dstPrefixes {
+		var (
+			srcPrefix *sourcePrefixes
+			match     int
+		)
+		switch srcType {
+		case SourceTypeExternal:
+			match = int(SourceTypeExternal)
+			srcPrefix = prefix.srcTypeArr[match]
+		case SourceTypeSameOrLoopback:
+			match = int(SourceTypeSameOrLoopback)
+			srcPrefix = prefix.srcTypeArr[match]
+		}
+		if srcPrefix == nil {
+			match = int(SourceTypeAny)
+			srcPrefix = prefix.srcTypeArr[match]
+		}
+		if match < bestSrcTypeMatch {
+			continue
+		}
+		if match > bestSrcTypeMatch {
+			bestSrcTypeMatch = match
+			srcPrefixes = make([]*sourcePrefixes, 0)
+		}
+		if srcPrefix != nil {
+			// The source type array always has 3 entries, but these could be
+			// nil if the appropriate source type match was not specified.
+			srcPrefixes = append(srcPrefixes, srcPrefix)
+		}
+	}
+	return srcPrefixes
+}
+
+// filterBasedOnSourcePrefixes is the third stage of the filter chain matching
+// algorithm. It trims the filter chains based on the source prefix. Only filter
+// chains with the most specific match progress to the next stage.
+func filterBasedOnSourcePrefixes(srcPrefixes []*sourcePrefixes, srcAddr net.IP) []*sourcePrefixEntry {
+	var (
+		matchingSrcPrefixes []*sourcePrefixEntry
+		maxSubnetMatch      int
+	)
+	for _, sp := range srcPrefixes {
+		for _, prefix := range sp.srcPrefixes {
+			if !prefix.net.Contains(srcAddr) {
+				continue
+			}
+			matchSize, _ := prefix.net.Mask.Size()
+			if matchSize < maxSubnetMatch {
+				continue
+			}
+			if matchSize > maxSubnetMatch {
+				maxSubnetMatch = matchSize
+				matchingSrcPrefixes = make([]*sourcePrefixEntry, 0)
+			}
+			matchingSrcPrefixes = append(matchingSrcPrefixes, prefix)
+		}
+	}
+	return matchingSrcPrefixes
+}
+
+// filterBasedOnSourcePorts is the last state of the filter chain matching
+// algorithm. It trims the filter chains based on the source ports. It expects
+// to be left with a single matching filter chain and returns an error if there
+// are multiple matching filter chains at the end.
+func filterBasedOnSourcePorts(srcPrefixEntries []*sourcePrefixEntry, srcPort int) (*FilterChain, error) {
+	// We need to find the most specific match from each of these source prefix
+	// entries. Matching filter chains are associated with a weight of 0 or 1,
+	// indicating whether they were a wildcard or a specific port match
+	// respectively. Once all source prefix entries have been processed, if we
+	// are left with more than one matching filter chain at the highest weight,
+	// it means that we have a match conflict.
+	matchingFCs := make(map[int][]*FilterChain)
```

Comment:
Would this be better as two different slices of *FilterChains instead of a map where 0 and 1 mean special things?

Reply:
Went with two vars (one a slice, and the second a pointer to a FilterChain). Thanks.